markdown/ch1.markdown

```diff
@@ -554,34 +581,38 @@ def class_boxplot(data, classes, colors=None, **kwargs):
     kwargs : dict
         Keyword arguments to pass on to `plt.boxplot`.
     """
-    # default color palette
+    # default boxplot parameters; only updated if not specified
+    kwargs.setdefault('sym', '.')
+    kwargs.setdefault('whiskerprops', {'linestyle': '-'})
+
+    # Set default color palette and map classes to colors
     if colors is None:
         colors = sns.xkcd_palette(["windows blue", "amber", "greyish",
                                    "faded green", "dusty purple"])
-    # default boxplot parameters; only updated if not specified
-    kwargs['sym'] = kwargs.get('sym', '.')
-    kwargs['whiskerprops'] = kwargs.get('whiskerprops', {'linestyle': '-'})
-
     all_classes = sorted(set(classes))
     class2color = dict(zip(all_classes, it.cycle(colors)))
-    # create a dictionary containing data of same length but only data
-    # from that class
-    class2data = {}
-    for i, (distrib, cls) in enumerate(zip(data, classes)):
+
+    # map classes to data vectors
+    # other classes get an empty list at that position for offset
+    class2data = defaultdict(list)
+    for distrib, cls in zip(data, classes):
         for c in all_classes:
-            class2data.setdefault(c, []).append([])  # empty dataset at first
+            class2data[c].append([])
```

Comment:
It has to do with the way the boxplot is drawn. Each class needs to draw an "empty" box plot where the other classes will be plotted. I couldn't find a better way of doing this.